File: infinitetrading/src/api/gateway/endpoints/compoundV3.R
# Compound V3 (Comet) sub-router
# Endpoints: POST /compoundV3/lend  POST /compoundV3/unlend
# Sanitisation: basic_check() validates apiKey, network (from DB cache), pool address.
#               isValidEthereumAddress() validates the asset token address.
# Forwarding:   Calls Express API (ep = localhost:8000) depositCompoundV3 / withdrawCompoundV3.

compoundV3 <- Plumber$new()

compoundV3$handle("POST", "/lend",
  function(apiKey, protocol = "dhedge", pool, network, asset, share = 100, amount = NULL) {
    protocol <- tolower(trimws(protocol))
    pool     <- tolower(trimws(pool))
    network  <- tolower(trimws(network))

    check <- basic_check(network = network, protocol = protocol, pool = pool, apiKey = apiKey)
    if (check$status == "fail") return(check)

    if (!isValidEthereumAddress(asset))
      return(list(status = "fail", status_code = 1004, message = "Invalid asset address"))
    asset <- tolower(asset)

    url <- paste0(ep, "depositCompoundV3",
                  "?apiKey=",  apiKey,
                  "&network=", network,
                  "&pool=",    pool,
                  "&asset=",   asset)

    share_num  <- suppressWarnings(as.numeric(share))
    amount_num <- suppressWarnings(as.numeric(amount))

    if (!is.null(amount) && !is.na(amount_num) && is.finite(amount_num) && amount_num > 0) {
      url <- paste0(url, "&amount=", round(amount_num, 6))
    } else if (!is.null(share) && !is.na(share_num) && is.finite(share_num) &&
               share_num > 0 && share_num <= 100) {
      url <- paste0(url, "&share=", round(share_num, 2))
    } else {
      return(list(status = "fail", status_code = 1009,
                  message = "Please specify share (1-100) or amount (>0)"))
    }

    response <- POST(url)
    txt <- tryCatch(content(response, "text", encoding = "UTF-8"), error = function(e) "N/A")
    if (status_code(response) == 200) {
      parsed <- tryCatch(jsonlite::fromJSON(txt), error = function(e) NULL)
      return(if (!is.null(parsed)) parsed else list(status = "success", msg = txt))
    }
    return(list(status = "fail", status_code = status_code(response),
                message = tryCatch(jsonlite::fromJSON(txt), error = function(e) txt)))
  },
  comment = "Supply an asset to Compound V3 (Comet) within a dHEDGE vault. Specify share (% of vault asset balance, 1-100) or a fixed amount."
)

compoundV3$handle("POST", "/unlend",
  function(apiKey, protocol = "dhedge", pool, network, asset, share = 100, amount = NULL) {
    protocol <- tolower(trimws(protocol))
    pool     <- tolower(trimws(pool))
    network  <- tolower(trimws(network))

    check <- basic_check(network = network, protocol = protocol, pool = pool, apiKey = apiKey)
    if (check$status == "fail") return(check)

    if (!isValidEthereumAddress(asset))
      return(list(status = "fail", status_code = 1004, message = "Invalid asset address"))
    asset <- tolower(asset)

    url <- paste0(ep, "withdrawCompoundV3",
                  "?apiKey=",  apiKey,
                  "&network=", network,
                  "&pool=",    pool,
                  "&asset=",   asset)

    share_num  <- suppressWarnings(as.numeric(share))
    amount_num <- suppressWarnings(as.numeric(amount))

    if (!is.null(amount) && !is.na(amount_num) && is.finite(amount_num) && amount_num > 0) {
      url <- paste0(url, "&amount=", round(amount_num, 6))
    } else if (!is.null(share) && !is.na(share_num) && is.finite(share_num) &&
               share_num > 0 && share_num <= 100) {
      url <- paste0(url, "&share=", round(share_num, 2))
    } else {
      return(list(status = "fail", status_code = 1009,
                  message = "Please specify share (1-100) or amount (>0)"))
    }

    response <- POST(url)
    txt <- tryCatch(content(response, "text", encoding = "UTF-8"), error = function(e) "N/A")
    if (status_code(response) == 200) {
      parsed <- tryCatch(jsonlite::fromJSON(txt), error = function(e) NULL)
      return(if (!is.null(parsed)) parsed else list(status = "success", msg = txt))
    }
    return(list(status = "fail", status_code = status_code(response),
                message = tryCatch(jsonlite::fromJSON(txt), error = function(e) txt)))
  },
  comment = "Withdraw (unlend) an asset from Compound V3 (Comet) within a dHEDGE vault. Specify share (% of cToken balance, 1-100) or a fixed amount."
)

# mount onto the main router (relies on `pr` being in scope when sourced)
pr$mount("/compoundV3", compoundV3)

File: infinitetrading_api/express/src/requests/compoundV3.ts
/**
 * requests/compoundV3.ts — Node port of R gateway's compoundV3 sub-router
 * (src/api/gateway/endpoints/compoundV3.R), mounted at /compoundV3/* in the
 * R gateway (port 8003). Historically: /compoundV3/lend & /compoundV3/unlend
 * -> basic_check() -> Express's own /depositCompoundV3 & /withdrawCompoundV3
 * directly (port 8000, no port-8002 involvement at all — these were never
 * trader-gated in R, replicated faithfully; see aaveV3.ts's header comment
 * for the broader auth-strength discussion across these three sub-routers).
 *
 * PARITY NOTES:
 *  - basic_check() only (network/protocol/apiKey/pool format), no
 *    isValidTrader.
 *  - asset must be a valid ethereum address (fail -> status_code 1004),
 *    lower-cased.
 *  - amount takes precedence over share if amount is numeric and > 0
 *    (rounded to 6 decimals); else share must be numeric in (0,100] (rounded
 *    to 2 decimals); else fail -> status_code 1009.
 *  - Forwards directly to Express's existing /depositCompoundV3 or
 *    /withdrawCompoundV3 (NOT renamed — these were never proxied through a
 *    conflicting public route name, so no *Raw rename was needed for them).
 */

import { Router, Request, Response } from 'express';
import { basicCheck, toRWireFormat, isValidEthereumAddress } from '../basicCheck';

const router = Router();

const EXPRESS_BASE = process.env.EXPRESS_INTERNAL_URL || 'http://localhost:8000/';

async function proxyPost(url: string): Promise<any> {
  const resp = await fetch(url, { method: 'POST' });
  const data: any = await resp.json().catch(() => ({}));
  if (resp.status === 200) return data;
  return { status: 'fail', status_code: resp.status, message: data };
}

function buildShareAmountSuffix(q: any): { suffix: string; error?: { status: string; status_code: number; message: string } } {
  const shareRaw = q.share === undefined ? '100' : String(q.share);
  const amountRaw = q.amount === undefined ? undefined : String(q.amount);
  const shareNum = Number(shareRaw);
  const amountNum = amountRaw === undefined ? NaN : Number(amountRaw);

  if (amountRaw !== undefined && !Number.isNaN(amountNum) && Number.isFinite(amountNum) && amountNum > 0) {
    return { suffix: `&amount=${Math.round(amountNum * 1e6) / 1e6}` };
  }
  if (q.share !== undefined && !Number.isNaN(shareNum) && Number.isFinite(shareNum) && shareNum > 0 && shareNum <= 100) {
    return { suffix: `&share=${Math.round(shareNum * 100) / 100}` };
  }
  return { suffix: '', error: { status: 'fail', status_code: 1009, message: 'Please specify share (1-100) or amount (>0)' } };
}

async function handleLendOrUnlend(req: Request, res: Response, endpoint: 'depositCompoundV3' | 'withdrawCompoundV3') {
  const q = { ...req.query, ...req.body };
  const protocol = String(q.protocol || 'dhedge').toLowerCase();
  const pool = String(q.pool || '').toLowerCase();
  const network = String(q.network || '').toLowerCase();
  const apiKey = String(q.apiKey || '');
  let asset = String(q.asset || '');

  const check = await basicCheck({ network, protocol, pool, apiKey });
  if (check.status === 'fail') return res.json(toRWireFormat(check));

  if (!isValidEthereumAddress(asset)) {
    return res.json({ status: 'fail', status_code: 1004, message: 'Invalid asset address' });
  }
  asset = asset.toLowerCase();

  let url = `${EXPRESS_BASE}${endpoint}?apiKey=${encodeURIComponent(apiKey)}&network=${network}&pool=${pool}&asset=${asset}`;
  const { suffix, error } = buildShareAmountSuffix(q);
  if (error) return res.json(error);
  url += suffix;

  return res.json(await proxyPost(url));
}

/**
 * @openapi
 * /compoundV3/lend:
 *   post:
 *     summary: (weak-auth sub-router) Supply an asset to Compound V3 (Comet) within a dHEDGE vault
 *     tags: [Lending]
 */
router.post('/compoundV3/lend', (req, res) => handleLendOrUnlend(req, res, 'depositCompoundV3'));

/**
 * @openapi
 * /compoundV3/unlend:
 *   post:
 *     summary: (weak-auth sub-router) Withdraw an asset from Compound V3 (Comet) within a dHEDGE vault
 *     tags: [Lending]
 */
router.post('/compoundV3/unlend', (req, res) => handleLendOrUnlend(req, res, 'withdrawCompoundV3'));

export default router;
